ContinousBenchmarking.md

```diff
@@ -0,0 +1,47 @@
+Continuous Benchmarking
+=======
+
+## Introduction:
+Benchmarking an algorithm is a compute intensive and time consuming chore that is essential for guaranteeing performance. 
+An algorithm can unknowingly drift towards an undesirable state if the laborious task of benchmarking is neglected for
+too long.  We envision a continous benchmarking system thats runs at least nightly, and posts results for the community to see.
+
+## Current State:
+Benchmarking is currently performed manually, by running a series of benchmarking scripts that can take up
+24 hours to complete.
+
+
+## Our current Tools:
+**Travis CI.** 
+    The current CI tool to manage the workflow of building garage and executing unit tests.  The Travis CI build 
+    environment have a maximum of 7.5 GB of memory, and 2 cores.  It also limits jobs to 120 minutes maximum for a
+    private Github repo, and 50 minutes for a public Github repo.  
+    https://docs.travis-ci.com/user/customizing-the-build/
+    https://docs.travis-ci.com/user/reference/overview/
+
+**Github Actions:**
+    Each workflow can run for up to 58 minutes (including queuing and execution time).  Two workflows can be running 
+    concurrently per repository.  1 virtual CPU.  Up to 3.75 GB of memory
+    https://developer.github.com/actions/creating-workflows/workflow-configuration-options/
+
+For Travis and Github actions, the compute limits are too restrictive to run the benchmarking in their environments.  
+Thus, we have applied for GCP and AWS credits to spin up cloud VMs to run the benchmark jobs.  The next question would 
+be whether to trigger these jobs from TravisCI/Github Actions, or to have a parallel CI system for managing the 
+benchmarking.
+I imagine we would have persistent VMs running, ready to receive the command to run the benchmarks.  The agent on the 
+VM would then pull down the appropriate garage branch, run the benchmarks, and send the results back to another persistent machine.
+
+## Proposed Tool:
+**Teamcity:**
```

Comment:
* We will have build this system to start by using machine(s) at USC at the build agents, so the local element will be important immediately.
* Being super efficient with cloud resources is nice-to-have but we can also just request enough credit to deal with our inefficiency. The numbers won't be significantly different from the cloud svcs perspective (e.g. it would make a 2x difference in spend, but not a 10x difference).
* You might want to check into CircleCI build time and resource limits. I know it's popular for bigger/more complicated build jobs that Travis can't support.

Can you provide a brief overview of alternative options and what implementation would look like using them?

I'm sure there are services available which could provide comparable features to TeamCity for our use case. We should also compare to the simplest home-grown solution we can come up with.

Here's an example of a minimal home-grown alternative to compare your design to:
* A head node (tiny VM) on a cloud provider has a cron job which runs every 24h
* The cron job downloads the latest garage master and ships it to (a) a USC machine for which it has SSH credentials (b) an AWS/GCP instance which it spins up.
* The build node leaves its benchmark logs/results in a cloud storage bucket or the head node just SCPs them into one (head node can just busy wait stdout on the SSH session or the build node can post an HTTP request with metadata back to the head node instead, or etc.)
* The head node renders the results to static HTML and updates a github.io page with them (alternatively it can update some DB tables if our display is DB-backed).
* Even simpler: no head node, just a cron job on a single machine at USC

Reply:
Kevin, can you elaborate a little bit on the open source license? Are there any limits we are likely to hit?

If the open source license really will be sufficient for us, I agree it looks like a very good option. It would take us a lot of work to build an interface as full-featured as what it provides.

Reply:
If the local element is going to kick start every CI build, then would we need to VPN into the local network to kick start it? Or will connectivity issues in the lab adversely affect our CI?

I'm not in favour of a homegrown solutions due to maintenance issues, especially when free CI options exist. Cron job on a single machine is definitely simple, but probably inefficient use of resources.

Reply:
@gautams3 Good question, both GCP and AWS has configs for automatically VPNing into our network. https://cloud.google.com/vpn/docs/concepts/overview

Reply:
Setting up cloud VPN is not for the faint of heart. RESL doesn't have an IPSEC VPN, anyway. 

This is why I suggested the SSH route. We could also setup a wireguard VPN manually between static machines in cloud and on-prem to achieve nearly the same effect.

Reply:
@ryanjulian Wireguard vpn does seem like a good option then.  

We would have to weigh the amount of work for creating a home grown solution versus the amount of work in setting up the vpn.  I believe teamcity/jenkins can already handle all of the minimal requirements you listed for the home grown option.

For cataloging results:
Teamcity has the option to save any output files as build artifacts.  Initially, we can store the outputted matplotlib graphs as build artifacts, and use the teamcity API to grab them for our results web page.  TC allows us to query for the latest successful build, so we can have a static html page hardcoded to always use this latest build.

If we decide to go with the homegrown option, I would suggest creating some db-backed web app instead.

Reply:
Can you include a proposal for display in you main proposal file?

Ideally this proposal should cover the entire end-to-end life of a benchmark (i.e. how do we get from a git hash to a benchmark plot on any browser screen).

I think right now it focuses on machine allocation and workflow management, which are problems we don't really have if we just use a single machine. All that is necessary for a minimum viable system is a single machine.

re: using a "custom" solution
I think it's important for everyone on this thread to take a step back and realize that this system will be custom no matter what, which is why we're bothering to even design it. The only OTS component under consideration could be replaced by a cron job with 0 change in user-facing functionality.

Well thought-out assemblies of proven, simple components are often far less work to maintain than even the most magical-seeming OTS products. Bending your design to match the API of  3rd-party product with a bunch of features you don't need is usually at least as much work as doing it yourself using just the basics. What the 3rd party product usually brings is some chrome and better scaling in some dimensions, which it's not clear we even need.

This is why I insist that we compare our design across 3rd party solutions and to the simplelist workable "first-party" solution. It helps us identify what the actual "custom" parts of the problem are, and helps us avoid bending the design to a single 3rd party API.

If we think that setting up a cron job on a box in the corner of the lab to do all this will be "less work" than figuring out an API integration with a cloud CI orchestration system + individual AWS or GCP APIs to work with it, then we either (a) have a very have a very strange definition of "work", or (b) are already completely hosed. If we can't make a shell script to do this and attach it to a cron job on some box, we certainly aren't going to get the fancy version right.

Reply:
@ryanjulian Should this proposal be agnostic of the tools we decide to use?   As it is not clear if we would scale beyond a single machine.  Perhaps I could only describe the exact purposes of the workflow that each tool is intended to fulfill.  I agree that setting up cloud ci system cloud be too much work if we are just going to use a single machine in the lab.  I only suggested such a design for the concerns of future scalability, with the assumption that we may have cloud computing credits to manage.

The basic workflow of detecting a VCS change, pulling that change to a test environment, running steps, logging, and storing results are common to the several CI tools I have seen, not just Teamcity.  Of course we could write a shell script to accomplish the same workflow, my only reservation is this may be re-inventing the wheel a little bit.

I guess the question is, how likely are we to scale from the running a single to a  few machines, or even to a cloud CI system.  I think this may help us answer how much investment on tools to make initially.

Reply:
the background should be angnostic but the design parts can be tool specific. once you establish the consensus on Background section, the requirements typically fall out on their own.

for instance multiple machines might be a feature/requirement, but implementing that is complicated. that feature needs to justify its existence by satisfying a goal or solving a problem mentioned in the background.